crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -264,12 +265,11 @@ async fn app_validation_workflow_inner(
             t,
             a,
             r,
-            total
+            ops_validated
         );
     }
     jh.await?;
-    tracing::debug!("accepted {} ops", total);
-    Ok(if saturated {
+    Ok(if saturated || ops_validated < num_ops_to_validate {
         WorkComplete::Incomplete
     } else {
         WorkComplete::Complete
```

Comment:
Won't this just endlessly retrigger? This may cause churn on this workflow running over and over as fast as possible until missing deps can be fetched, which could be 1000s of triggers.

Maybe `Incomplete` needs to take a Duration for delay. When saturated, the delay can be 0, and when just incomplete, it can be something more sensible.

@ThetaSinner was also suggesting we rework this whole pattern and just move the zero-delay loops into the workflow itself, and use delays when needing to retrigger, so what I just suggested might be a step in that direction.

Reply:
Well no, not just endlessly retrigger unless validation always fails. Sure, a delay would slow this down, but my question is whether we can find a better mode of retriggering altogether. Because with or without a delay, the retrigger is based on hoping that things will go better this time around.

Reply:
Yes, the ideal would be to correctly identify all the reasons why we should re-run validation and only retrigger when those events occur, but without pubsub we can't get there all way, because one reason to fail and re-run validation is because some validation dependency that some other node is holding is now available and wasn't before, which we can't know without attempting to fetch the dep. Seems like the only way to know for sure if things will go better next time around is to re-run validation and find out if things went better.

A reasonable delay in retriggering will give us polling of validation dependencies in lieu of proper pubsub.

Reply:
What I was saying about `Incomplete` was slightly stronger https://github.com/holochain/holochain/pull/2922#discussion_r1352809001

I think it's a recipe for infinite loops that we don't actually need in any case I've seen yet. I think the point about state changing on a peer allowing us to proceed with validation is a good motivation for this. Why go into a loop when we have a missing dependency without solid mechanisms to break us out of that loop, and solid tests to prove they work.

I think the approach that the publish workflow takes is more sensible. It checks for work, if there's no work then it suspends. If there is work then it attempts that work. Anything that was touched gets put on a timeout so that it won't be visible as work to be done for a while. When new work arrives and the workflow is triggered then it resumes its looping until that work is done.

I think that's really simple logic to follow now that it's the only logic and that workflow doesn't use `Incomplete` any more.

Whereas `Incomplete` is complex to follow. It's potentially returned from several places in the large app/sys validation files and although it isn't actually returned from too many places I cannot decide its correctness by reading alone. Using it to create a temporary loop when what you actually mean is "I processed this set of things out of order and I need to run again" just seems like a problem that the workflow code should solve.

I think the thing to focus on is writing tests for how the workflow handles inputs in different orders and once we're sure it does the best it can with a single run, then think about triggers. In my understanding, if we needed an op from a peer and we now have that op then the op should be sitting in our fetch queue. It's arrival should be a valid trigger to kick app validation again and see if any more ops can proceed.

I think a loop could work for validation like it does for publish but explicit triggers would be preferable. Either way, there's a lot of untested code here that needs pulling apart before I would want to make too many changes to it.

Sorry for the essay! Mostly thinking out loud (typed) :)

Reply:
I'm coming back to this after our app validation design session. We have decided to trigger app validation periodically as a quick fix while working on a larger refactor of it.

What that comes down to is to trigger this workflow again after it has run, after a delay. This is essentially what this PR does, with the added condition that the re-triggering only happens if not all ops that are in the DB to be validated have been validated. So it is more restrictive than simply a permanent delayed loop of runs. It should therefore potentially cause even less trouble than the unconditional triggering.

@maackle @ThetaSinner @thedavidmeister Up for your review again, please.